core-backend/common/src/lib.rs

```diff
@@ -130,34 +133,47 @@ pub trait Environment<E: Ext + IntoExtInfo + 'static>: Sized {
     fn into_gas_amount(self) -> GasAmount;
 }
 
-pub trait OnSuccessCode<T, E> {
-    fn on_success_code<F>(self, f: F) -> Result<i32, E>
-    where
-        F: FnMut(T) -> Result<(), E>;
+pub struct ExtErrorProcessor<'a, T, E> {
+    ext: &'a mut E,
+    success: Option<T>,
 }
 
-impl<T, E, E2> OnSuccessCode<T, E> for Result<T, E2> {
-    fn on_success_code<F>(self, mut f: F) -> Result<i32, E>
+impl<'a, T, E> ExtErrorProcessor<'a, T, E>
+where
+    E: Ext + IntoExtInfo,
+{
+    pub fn new(ext: &'a mut E) -> Self {
+        Self { ext, success: None }
+    }
+
+    pub fn with<U, F>(mut self, f: F) -> Result<Self, U>
```

Comment:
I'm not a fan of such kind functions with side effects like 'with' and 'on_success' because they have a quite strict using contract and may yield errors when called improperly. May we pass 'Result' instead of 'Self' to make them more pure?

Reply:
We cannot pass `T` instead because `Self` is used further. If you afreaid of robustness I can use compile-time state so contract won't be violated

Reply:
As I see it is used for `error_len()` only. Ok, if there is no better solution, some compile-time checks would be great.